spec/controllers/spree/admin/payments_controller_spec.rb

```diff
@@ -66,5 +66,59 @@
         end
       end
     end
+
+    context "requesting a partial credit on a payment" do
+      let(:params) { { id: payment.id, order_id: order.number, e: :credit } }
+
+      # Required for the respond override in the controller decorator to work
+      before { @request.env['HTTP_REFERER'] = spree.admin_order_payments_url(payment) }
+
+      context "that was processed by stripe" do
+        let!(:payment_method) { create(:stripe_payment_method, distributors: [shop], preferred_enterprise_id: shop.id) }
+        let!(:payment) { create(:payment, order: order, state: 'completed', payment_method: payment_method, response_code: 'ch_1a2b3c', amount: order.total + 5) }
+
+
+        before do
+          allow(Stripe).to receive(:api_key) { "sk_test_12345" }
+        end
+
+        context "where the request succeeds" do
+          before do
+            stub_request(:post, "https://sk_test_12345:@api.stripe.com/v1/charges/ch_1a2b3c/refunds").
+              to_return(:status => 200, :body => JSON.generate(id: 're_123', object: 'refund', status: 'succeeded') )
+          end
+
+          it "partially refunds the payment" do
+            order.reload
+            expect(order.payment_total).to eq order.total + 5
+            expect(order.outstanding_balance).to eq(-5)
+            spree_put :fire, params
+            expect(payment.reload.state).to eq 'completed'
+            order.reload
+            expect(order.payment_total).to eq order.total
+            expect(order.outstanding_balance).to eq 0
+          end
```

Comment:
I find confusing the fact that there are two `expect` blocks (verification phase) in this test. I don't know whether the subject of the test is `spree_put :fire, params` or something else I don't see. IMO tests are easier to read when using the [four-phase approach](https://robots.thoughtbot.com/four-phase-test): setup, exercise, verify and teardown.

Reply:
Are you suggesting that we shouldn't make more than one assertion per spec? Would you be happier if I made this a request spec?

Reply:
@sauloperez I want to improve this, but I'm not really sure how. I want to test that all of these things happen, as they are all affects that I am expecting to see. This is very much an integration test, so it is checking that a bunch of different things are working properly together. Would you prefer if I had each of these expectations in their own example?

Reply:
> Would you prefer if I had each of these expectations in their own example?

Yes. You're actually asserting different things such as the fact that partially refunding the payment resets the outstanding balance, for instance.